crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -932,12 +935,59 @@ impl IndexerReader {
                 );
                 (inner_query, "1 = 1".into())
             }
-            Some(
-                TransactionFilter::TransactionKind(_) | TransactionFilter::TransactionKindIn(_),
-            ) => {
-                return Err(IndexerError::NotSupported(
-                    "TransactionKind filter is not supported.".into(),
-                ));
+            Some(TransactionFilter::TransactionKind(kind)) => {
+                if kind == IotaTransactionKind::SystemTransaction {
+                    ("tx_kinds".into(), "tx_kind = 0 OR tx_kind > 1".to_string())
+                } else {
+                    ("tx_kinds".into(), format!("tx_kind = {}", kind as u8))
+                }
+            }
+            Some(TransactionFilter::TransactionKindIn(kind_vec)) => {
+                if kind_vec.is_empty() {
+                    return Err(IndexerError::InvalidArgument(
+                        "no transaction kind provided".into(),
+                    ));
+                }
+
+                let mut has_system_transaction = false;
+                let mut has_programmable_transaction = false;
+                let mut other_kinds = Vec::new();
+
+                for kind in kind_vec.iter() {
+                    match kind {
+                        IotaTransactionKind::SystemTransaction => has_system_transaction = true,
+                        IotaTransactionKind::ProgrammableTransaction => {
+                            has_programmable_transaction = true
+                        }
+                        _ => other_kinds.push(*kind as u8),
+                    }
+                }
+
+                let query = if has_system_transaction {
+                    if has_programmable_transaction {
+                        // Case: Both `0` (SystemTransaction) and `1` (ProgrammableTransaction) are
+                        // included: Allow everything
+                        "tx_kind >= 0".to_string()
```

Comment:
is the filter even needed if we allow everything anyways?

Reply:
It's not really needed, refactored with an empty `String` in: https://github.com/iotaledger/iota/pull/5801/commits/9e31a7bdb5047d1513b53116e88e40ec027d6b30